storage-proofs/src/rational_post.rs

```diff
@@ -0,0 +1,461 @@
+    fn identifier(&self) -> String {
+        format!(
+            "rationalPoSt::PublicParams{{sector_size: {} challenges_count: {}}}",
+            self.sector_size(),
+            self.challenges_count,
+        )
+    }
+
+    fn sector_size(&self) -> u64 {
+        self.sector_size
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PublicInputs<'a, T: 'a + Domain> {
+    /// The challenges, which leafs to prove.
+    pub challenges: &'a [u64],
+    pub faults: &'a [u64],
+    /// The root hashes of the underlying merkle trees.
+    pub commitments: &'a [T],
+}
+
+#[derive(Debug, Clone)]
+pub struct PrivateInputs<'a, H: 'a + Hasher> {
+    pub trees: &'a [&'a MerkleTree<H::Domain, H::Function>],
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Proof<H: Hasher> {
+    #[serde(bound(
+        serialize = "MerkleProof<H>: Serialize",
+        deserialize = "MerkleProof<H>: Deserialize<'de>"
+    ))]
+    inclusion_proofs: Vec<MerkleProof<H>>,
+}
+
+impl<H: Hasher> Proof<H> {
+    pub fn leafs(&self) -> Vec<&H::Domain> {
+        self.inclusion_proofs
+            .iter()
+            .map(MerkleProof::leaf)
+            .collect()
+    }
+
+    pub fn commitments(&self) -> Vec<&H::Domain> {
+        self.inclusion_proofs
+            .iter()
+            .map(MerkleProof::root)
+            .collect()
+    }
+
+    pub fn paths(&self) -> Vec<&Vec<(H::Domain, bool)>> {
+        self.inclusion_proofs
+            .iter()
+            .map(MerkleProof::path)
+            .collect()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RationalPoSt<'a, H>
+where
+    H: 'a + Hasher,
+{
+    _h: PhantomData<&'a H>,
+}
+
+impl<'a, H: 'a + Hasher> ProofScheme<'a> for RationalPoSt<'a, H> {
+    type PublicParams = PublicParams;
+    type SetupParams = SetupParams;
+    type PublicInputs = PublicInputs<'a, H::Domain>;
+    type PrivateInputs = PrivateInputs<'a, H>;
+    type Proof = Proof<H>;
+    type Requirements = NoRequirements;
+
+    fn setup(sp: &Self::SetupParams) -> Result<Self::PublicParams> {
+        Ok(PublicParams {
+            sector_size: sp.sector_size,
+            challenges_count: sp.challenges_count,
+        })
+    }
+
+    fn prove<'b>(
+        pub_params: &'b Self::PublicParams,
+        pub_inputs: &'b Self::PublicInputs,
+        priv_inputs: &'b Self::PrivateInputs,
+    ) -> Result<Self::Proof> {
+        let sector_size = pub_params.sector_size;
+        let sector_count = priv_inputs.trees.len() as u64;
+
+        assert_eq!(
+            pub_inputs.challenges.len(),
+            pub_inputs.commitments.len(),
+            "missmatching challenges and commitments"
+        );
+        let challenges = pub_inputs.challenges;
+
+        let proofs = challenges
+            .iter()
+            .zip(pub_inputs.commitments.iter())
+            .map(|(challenge, commitment)| {
+                let challenged_sector = *challenge % sector_count;
+                let challenged_leaf = *challenge % (sector_size / 32);
```

Comment:
See https://github.com/filecoin-project/specs/pull/332/files#r308409694.

Given the complexity noted in the issue, I'd like to see derivation of challenged sectors and challenged leaves abstracted and shared between proof and verification — in such a way that we can carefully examine the method in one place to feel confident its safe.